bluemira/magnetostatics/greens.py

```diff
@@ -288,12 +278,44 @@ def greens_dpsi_dz(
 
     The implementation used here refactors the above to avoid some zero divisions.
     """
-    a = ((x + xc) ** 2 + (z - zc) ** 2) ** 0.5
+    a, k2 = calc_a_k2(xc, zc, x, z)
+    i1, i2 = calc_i1_i2(a, k2)
+    return MU_0_2PI * ((z - zc) * (i1 - i2 * ((z - zc) ** 2 + x**2 + xc**2)))
+
+
+@nb.jit(nopython=True)
+def calc_a_k2(
+    xc: float | np.ndarray,
+    zc: float | np.ndarray,
+    x: float | np.ndarray,
+    z: float | np.ndarray,
+):
+    a = np.hypot((x + xc), (z - zc))
     k2 = 4 * x * xc / a**2
+    # Avoid NaN when coil on grid point
     k2 = clip_nb(k2, GREENS_ZERO, 1.0 - GREENS_ZERO)
-    i1 = ellipk_nb(k2) / a
-    i2 = ellipe_nb(k2) / (a**3 * (1 - k2))
-    return MU_0_2PI * ((z - zc) * (i1 - i2 * ((z - zc) ** 2 + x**2 + xc**2)))
+    return a, k2
+
+
+@nb.jit(nopython=True)
+def calc_e_k(
+    k2: float | np.ndarray,
+):
+    return ellipe_nb(k2), ellipk_nb(k2)
+
+
+@nb.jit(nopython=True)
+def calc_i1_i2(
+    a: float | np.ndarray,
+    k2: float | np.ndarray,
+    e: float | np.ndarray | None = None,
+    k: float | np.ndarray | None = None,
+):
+    if (e is None) or (k is None):
+        e, k = calc_e_k(k2)
+    i1 = k / a
+    i2 = e / (a**3 * (1 - k2))
```

Comment:
Missed during the conversion
```suggestion
    i1 = 4 * k / a
    i2 = 4 * e / (a**3 * (1 - k2))
```

Reply:
actually sometimes its there an sometimes its not

Reply:
scratch this one